src/pdc-api.ts

```diff
@@ -50,6 +50,102 @@ const usePdcApi = <T>(
   return response;
 };
 
+const usePdcCallbackApi = <T>(
+  path: string,
+): (options: RequestInit) => Promise<T> => {
+  const { fetch } = useOidcFetch();
+
+  return useCallback((options: RequestInit) => {
+    const url = new URL(path, API_URL);
+    return fetch(url, options)
+      .then(throwNotOk)
+      .then((res) => res.json())
+      .catch((e) => logError(e, path, options));
+
+    /* eslint-disable-next-line react-hooks/exhaustive-deps --
+     *
+     * fetch should not be a dependency, because although it or its internal
+     * state may change from render to render, such changes are not relevant:
+     * a change in the way we make a request should not trigger an API request
+     */
+  }, [path]);
+};
+
+interface ApiPresignedPostRequest {
+  fileType: string;
+  fileSize: number;
+}
+
+interface ApiPresignedPostResponse {
+  fileType: string;
+  fileSize: number;
+  presignedPost: {
+    url: string;
+    fields: Record<string, string> & {
+      key: string,
+    };
+  };
+}
+
+const usePresignedPostCallback = () => {
+  const api = usePdcCallbackApi<ApiPresignedPostResponse>('/presignedPostRequests');
+  return (params: ApiPresignedPostRequest) => api({
+    method: 'post',
+    headers: {
+      Accept: 'application/json',
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify(params),
+  });
+};
+
+const uploadPresignedPost = (
+  post: ApiPresignedPostResponse,
+  file: File,
+) => {
+  const formData = new FormData();
+  Object.entries(post.presignedPost.fields).forEach(([key, value]) => (
+    formData.append(key, value)
+  ));
+  formData.append('Content-Type', post.fileType);
```

Comment:
@slifty It strikes me that we're mildly missing an opportunity to verify file type here. We're setting content type to the presigned post's `fileType`, but shouldn't we be using the file's type instead, just to double-verify that the file matches the presigned post?

In that case, this should read like so instead:

```suggestion
  formData.append('Content-Type', file.type || 'application/octet-stream');
```

Thoughts?

Reply:
I like it, but don't like having the "null type is overridden" in two places.
I'll probably go with something like `file.type || post.fileType`

Reply:
Nevermind -- I don't even want to have access to `post.fileType` (I only want the presignedPost rather than the presignedPost request result, so I've refactored that function).  I'll just double up the `||` for now.

Reply:
That's a good concern. However what do you think about just making a constant out of the fallback type and referencing it both places that way? I do think following the same logic here as for the presigned-post-generation seems important, otherwise there's essentially no file type validation if the second file (used for upload) has a `null` type.

Reply:
@slifty Ugh I keep replying to old things because GH's interface isn't pulling in latest replies fast enough.